### PR TITLE
HOLD - CLOUDSTACK-10132: Multiple Management Servers Support for agents

### PR DESCRIPTION
**Pull request by nvazquez**
JIRA Ticket: https://issues.apache.org/jira/browse/CLOUDSTACK-10132

Functional Specification: https://cwiki.apache.org/confluence/display/CLOUDSTACK/Multiple+Management+Servers+Support+on+agents

This feature allows the direct agents to support multiple management servers. With this feature, a load balancer may no longer be necessary. The CloudStack administrator is responsible for setting a list of management servers and an algorithm in which to sort them, to the management server, using global configurations. The management server is responsible for applying algorithm to the management server list and propagate it to the agents. 

### Overview
A new global configuration is added: 'direct.agent.lb.algorithm'. Possible values of this global configuration are: 'static', 'roundrobin' and 'shuffle', corresponding to the algorithm that would be applied to the list of comma separated hosts defined on 'host' global configuration before sending them to the agents.
- 'static': No modification is made to the original 'host' list
- 'roundrobin': Rotate items one by one once the list is read
- 'shuffle': Randomly sort the list

A central utility is reading values from both global configurations on management server startup and is responsible of propagating list to each agent. In propagated list, the first element acts as the main host. Load balancing takes place by prepending the management server each agent should connect to, to the propagated actual list.

Once agents connect to management server, they receive a comma separated string of management servers. This value is compared to their actual 'host' configuration, and in case they differ, 'host' property is updated and persisted on 'agent.properties' file with the newly received string.

Reconnection logic between agent and management server is refactored as well. If agent is not able to establish a connection to its actual management server after some timeout interval, it attempts connection to the next element on its 'host' list.

Also, a background task is added to the agents to attempt reconnection to their main management server in pursuit of balancing. The interval of this background task is set on 'hosts.lb.background.interval' property on 'agent.properties' file. It can be optionally disabled, by setting property to 0

### Example
Supposing an environment in which there are 3 management servers: A, B and C and 3 KVM agents.
Setting 'host' = 'A,B,C', agents will receive lists depending on 'direct.agent.lb' value:
- 'static': Each agent will receive the list: 'A,B,C'
- 'roundrobin': First agent receives: 'A,B,C', second agent receives: 'B,C,A', third agent receives: 'C,B,A'
- 'shuffle': Each agent will receive a list in random order.

### CODE REVIEW

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@nvazquez thanks for this PR, I'll run marvin tests

**Comment by borisstoyanov** [comment]
@blueorangutan test

File: agent/src/com/cloud/agent/AgentShell.java [hunk 1]
```diff
@@ -107,18 +109,37 @@ public String getPod() {
     }
 
     @Override
-    public String getHost() {
+    public String getNextHost() {
         final String[] hosts = _host.split(",");
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/AgentShell.java]: We can use getHosts() here.

File: agent/src/com/cloud/agent/AgentShell.java [hunk 2]
```diff
@@ -173,6 +194,19 @@ public void setPersistentProperty(String prefix, String name, String value) {
             _storage.persist(name, value);
     }
 
+    @Override
+    public boolean setPropertyAndPersist(String property, String value) {
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/AgentShell.java]: Don't use it, there is another way to do that --> call setPersistentProperty, see the method is just above your implementation.

File: utils/src/main/java/com/cloud/utils/PropertiesUtil.java [hunk 3]
```diff
@@ -193,4 +195,19 @@ public static Properties loadFromFile(final File file)
         return properties;
     }
 
+    public static void persist(Properties properties, File file, String key, String value) throws IOException {
```
Comment (rhtyd) [inline utils/src/main/java/com/cloud/utils/PropertiesUtil.java]: This can be removed as well. See how the keystore passphrase is persisted, follow that pattern. Hope this helps @nvazquez

**Comment by nvazquez** [comment]
Thanks @rhtyd for pointing out pattern for persistent properties, I've addressed your comments

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@nvazquez @borisstoyanov I've made and pushed some additional edge cases and bugfixes now. In my test env it seemed okay. Please continue reviewing and testing.
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@rhtyd @borisstoyanov please find new additions, kindly review/test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

File: agent/src/com/cloud/agent/Agent.java [hunk 4]
```diff
@@ -699,16 +719,97 @@ public void processResponse(final Response response, final Link link) {
...
         if (ready.getHostId() != null) {
             setId(ready.getHostId());
         }
-        s_logger.info("Ready command is processed: agent id = " + getId());
 
+        final List<String> msList = ready.getMsList();
+        if (msList != null && msList.size() > 0) {
+            final List<String> hosts = Arrays.asList(_shell.getHosts());
+            s_logger.info("Received management servers list: " + msList + ", current list: " + hosts);
+            if (isReceivedListUpdated(hosts, msList)) {
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: It's debatable if we want to update the `host` list when order has changed, I think if the hosts are same (but order is different) we should ignore. In my testing I was able to reproduce an infinite switching behaviour when preferred host (the first in list) was not the one the agent was connected to leading to period disconnections (with just two hosts in the list).

File: agent/src/com/cloud/agent/Agent.java [hunk 5]
```diff
@@ -699,16 +719,97 @@ public void processResponse(final Response response, final Link link) {
...
+                persistNewManagementServersList(msList);
+                if (!_shell.getConnectedHost().equals(msList.get(0))) {
+                    attemptReconnectionToPrimaryHost(msList.get(0));
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: Please remove the attempt to reconnect from here, the responsibiilty of doing that lies with the background task and not the ready command handler.

File: agent/src/com/cloud/agent/Agent.java [hunk 6]
```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
...
+                 final String[] msList = _shell.getHosts();
+                 String preferredHost  = msList[index];
+                 final String connectedHost = _shell.getConnectedHost();
+                 if (s_logger.isTraceEnabled()) {
+                     s_logger.trace("Running host LB checker task, connected host=" + connectedHost + ", preferred host=" + preferredHost);
+                 }
+                 final long wait = _shell.getFailbackPollingWait();
+                 while (!preferredHost.equals(connectedHost)) {
+                     if (_link != null) {
+                         boolean isHostUp = true;
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: Optional: The code to check if the host is reachable on a port/address can we refactored.

File: agent/src/com/cloud/agent/Agent.java [hunk 7]
```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
...
+             try {
+                 int index = 0;
+                 final String[] msList = _shell.getHosts();
+                 String preferredHost  = msList[index];
+                 final String connectedHost = _shell.getConnectedHost();
+                 if (s_logger.isTraceEnabled()) {
+                     s_logger.trace("Running host LB checker task, connected host=" + connectedHost + ", preferred host=" + preferredHost);
+                 }
+                 final long wait = _shell.getFailbackPollingWait();
+                 while (!preferredHost.equals(connectedHost)) {
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: Let's not have a while loop here. @nvazquez if you're not aware, when the agent is disconnected the `reconnect()` logic already kicks and is responsible for performing reconnections. This timer/task should not share the same responsibility, instead it should check if the preferred host (first in the list) is available and if the agent is already connected, only then it should disconnect by performing a reconnect on the `_link` (if it's not null for example).

Bear in mind, we've two assumptions here (1) the first host in the list is the preferred one, (2) the timer task is responsible for reconnecting to the preferred host.

File: agent/src/com/cloud/agent/Agent.java [hunk 8]
```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
...
+                             if (_inProgress.get() == 0) {
+                                 if (s_logger.isDebugEnabled()) {
+                                     s_logger.debug("Preferred host is found to be reachable, trying to reconnect to: " + preferredHost);
+                                 }
+                                 _shell.resetHostCounter();
+                                 reconnect(_link);
+                             }
+                         }
+                     }
+                     index = (index + 1) % msList.length;
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: @nvazquez don't change the preferred host to some other host, it's because we assume the first in the list is the `preferred` host. For example, consider this use-case when using the `static` algorithm the first item in the list could be IP of the LB (netscalar/haproxy etc) and the use of this mechanism is to ensure HA for the mgmt server when this LB is down. If we change the preferred host, it breaks our use-case.

File: agent/src/com/cloud/agent/Agent.java [hunk 9]
```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
...
+                             socket.connect(new InetSocketAddress(preferredHost, _shell.getPort()), 5000);
+                         } catch (final IOException e) {
+                             isHostUp = false;
+                             if (s_logger.isDebugEnabled()) {
+                                 s_logger.debug("Host: " + preferredHost + " is not reachable, trying to reconnect to next host...");
+                             }
+                         }
+                         if (isHostUp && _link != null) {
+                             long initialTime = System.currentTimeMillis();
+                             while (System.currentTimeMillis() - initialTime < wait && _inProgress.get() > 0) {
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: @nvazquez please avoid writing procedural code, in `reconnect` remaining tasks handling already exists. We should not have any timeouts/limits on these tasks, i.e. don't kill/purge outstanding tasks. Instead, one way to rewrite this could be to perform no operations when there are any tasks in `_inProgress`, and let the HostLBCheckerTask retry in the next round.

File: agent/conf/agent.properties [hunk 10]
```diff
@@ -30,6 +30,19 @@ workers=5
...
+# one) and will attempt to reconnect to the preferred host when it's connected
+# to one of the secondary/backup hosts.
+# The timer task is scheduled when agent starts for the first time, any change
+# in this setting requires restarting the agent.
+host.lb.interval=0
+
+# The time in seconds in which the agent will wait for active commands to be completed
+# before connecting to preferred host
+# Any change in this setting requires restarting the agent.
+failback.polling.wait=10
```
Comment (rhtyd) [inline agent/conf/agent.properties]: @nvazquez this can be removed, see my comments below.

File: agent/src/com/cloud/agent/Agent.java [hunk 11]
```diff
@@ -699,16 +719,97 @@ public void processResponse(final Response response, final Link link) {
...
+        _shell.setPersistentProperty(null, "host", newHosts);
+        _shell.resetHostCounter();
+        s_logger.info("Saved new management servers list: " + msList);
+    }
 
+    /**
+     * Try reconnecting to preferred management server after new list is received.
+     * If connection is not performed, host counter is reset
+     */
+    private void attemptReconnectionToPrimaryHost(String preferred) {
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: This can be removed, duplicates code too.

File: agent/src/com/cloud/agent/Agent.java [hunk 12]
```diff
@@ -699,16 +719,97 @@ public void processResponse(final Response response, final Link link) {
...
+     * Checks if received list is different to actual management server list (in order and size)
+     * @param actual actual list
+     * @param received received list
+     * @return true if list is updated, false if not
+     */
+    private boolean isReceivedListUpdated(List<String> actual, List<String> received) {
+        if (actual.size() != received.size()) {
+            return true;
+        }
+        for (int i = 0; i < received.size(); i++) {
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: We can simply using the previous implementation to simply check if the existing setting differs in size or `set`.

File: agent/src/com/cloud/agent/Agent.java [hunk 13]
```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
...
+                             while (System.currentTimeMillis() - initialTime < wait && _inProgress.get() > 0) {
+                                 s_logger.debug("Waiting for active commands to be completed");
+                                 Thread.sleep(5000l);
+                             }
+                             if (_inProgress.get() == 0) {
+                                 if (s_logger.isDebugEnabled()) {
+                                     s_logger.debug("Preferred host is found to be reachable, trying to reconnect to: " + preferredHost);
+                                 }
+                                 _shell.resetHostCounter();
+                                 reconnect(_link);
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: @nvazquez just so we're on the same page, when performing a reconnect after resting the host counter, `reconnect` will start getting host from the first in the list, i.e. the preferred one.

File: agent/src/com/cloud/agent/AgentShell.java [hunk 14]
```diff
@@ -107,18 +111,48 @@ public String getPod() {
...
+    public String getNextHost() {
+        final String[] hosts = getHosts();
         if (_hostCounter >= hosts.length) {
             _hostCounter = 0;
         }
-        final String host = hosts[_hostCounter % hosts.length];
+        connectedHost = hosts[_hostCounter % hosts.length];
         _hostCounter++;
-        return host;
+        s_logger.info("Connecting to host: " + connectedHost);
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/AgentShell.java]: This method only returns the next possible host in list, however, this does not mean we're connected to the host yet. This might cause threading issues, and in rare cases the connectedHost might not be the connected host. Reconsider how this could be re-written, perhaps the connectedHost can be updated using one of the `Task.CONNECT` handlers.

File: agent/src/com/cloud/agent/AgentShell.java [hunk 15]
```diff
@@ -291,6 +325,11 @@ protected boolean parseCommand(final String[] args) throws ConfigurationExceptio
             _properties.setProperty("guid", _guid);
         }
 
+        String val = getProperty(null, "host.lb.interval");
+        hostLBTimerInterval = (Strings.isNullOrEmpty(val) ? 0L : Long.valueOf(val));
+        String wait = getProperty(null, "failback.polling.wait");
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/AgentShell.java]: Remove this.

File: agent/src/com/cloud/agent/IAgentShell.java [hunk 16]
```diff
@@ -22,33 +22,64 @@
...
 
-    public String getVersion();
+    String getVersion();
+
+    void setHosts(String hosts);
+
+    /**
+     * Resets host counter to start from the beginning
+     */
+    void resetHostCounter();
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/IAgentShell.java]: Remove comments, or add them for all the methods on this interface for consistency.

File: server/src/com/cloud/resource/ResourceManagerImpl.java [hunk 17]
```diff
@@ -800,6 +804,7 @@ public Discoverer getMatchingDiscover(final Hypervisor.HypervisorType hypervisor
 
                 }
                 s_logger.info("server resources successfully discovered by " + discoverer.getName());
+                agentMSLB.propagateManagementServerListToAgents();
```
Comment (rhtyd) [inline server/src/com/cloud/resource/ResourceManagerImpl.java]: @nvazquez this is not desirable, this will cause the list to be sent to all the hosts whenever any single agent connects.

File: server/src/org/apache/cloudstack/agent/mslb/AgentMSLBServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+                    .collect(Collectors.toList());
+        }
+        return null;
+    }
+
+    @Override
+    public void propagateManagementServerListToAgents() {
+        List<HostVO> hosts = getOrderedRunningIndirectAgentIds();
+        if (hosts != null) {
+            for (HostVO h : hosts) {
```
Comment (rhtyd) [inline server/src/org/apache/cloudstack/agent/mslb/AgentMSLBServiceImpl.java]: @nvazquez I don't think it's a good idea, `resource claim and ownership` is not within the scope of this feature. This method will send command/answers to all the hosts (even those not managed by the `this` management server) and may cause disconnections. In case of multiple management servers, this will cause all the management servers to send multiple requests to all the agents. (note, when you do agent.send(), if the agent is not connected to `this` management server, it will use clustered mgmt server impl to send the request to the forwarding agent which is via other mgmt servers).

File: agent/src/com/cloud/agent/Agent.java (hunk 7 above)
Comment (nvazquez) [reply agent/src/com/cloud/agent/Agent.java]: Done, thanks

File: agent/src/com/cloud/agent/Agent.java (hunk 8 above)
Comment (nvazquez) [reply agent/src/com/cloud/agent/Agent.java]: Thanks @rhtyd, I assumed background task should attempt reconnection to any available host starting from the preferred one at the beggining of the list and trying in order on 'host' list. I refactored this background task to attempt reconnection only to preferred host as you suggested, I think it simplifies the feature

File: agent/src/com/cloud/agent/Agent.java (hunk 9 above)
Comment (nvazquez) [reply agent/src/com/cloud/agent/Agent.java]: Done, thanks! Agreed

File: agent/src/com/cloud/agent/Agent.java (hunk 13 above)
Comment (nvazquez) [reply agent/src/com/cloud/agent/Agent.java]: Right, my bad. Now it makes sense as timer task will attempt reconnection to the first host in the list

File: agent/src/com/cloud/agent/AgentShell.java (hunk 14 above)
Comment (nvazquez) [reply agent/src/com/cloud/agent/AgentShell.java]: Done, thanks! Please checkout `hostToConnect` and `connectedHost` relationship using `updateConnectedHost()` method on connection to a host.

**Comment by nvazquez** [comment]
@rhtyd thanks for your valuable review! I refactored code addressing most of your comments. There are quite few in which I think we should get aligned on and once we get aligned I could refactor as well! Thanks!

**Comment by borisstoyanov** [comment]
@blueorangutan package

File: agent/src/com/cloud/agent/Agent.java (hunk 4 above)
Comment (DaanHoogland) [reply agent/src/com/cloud/agent/Agent.java]: That seems to be a design issue then. When the order has changed the (ordered) list should be updated. It is debatable if the connection should be reset, though.

File: agent/src/com/cloud/agent/IAgentShell.java (hunk 16 above)
Comment (DaanHoogland) [reply agent/src/com/cloud/agent/IAgentShell.java]: I do not feel consistency is an issue here as much as the comment not giving more information than the methodname, so yes, please remove.

**Comment by borisstoyanov** [comment]
@blueorangutan package

File: agent/src/com/cloud/agent/Agent.java [hunk 19]
```diff
@@ -138,12 +143,14 @@ public int value() {
     private String _keystoreSetupPath;
     private String _keystoreCertImportPath;
 
+    private Long connectionTimeout = 120000L;
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: @nvazquez This is not used anywhere, can be removed.

File: agent/src/com/cloud/agent/Agent.java [hunk 20]
```diff
@@ -216,6 +223,7 @@ public String getPod() {
     }
 
     protected void setLink(final Link link) {
+        s_logger.info("Setting up link!");
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: I had put this for logging/debugging, may be removed. @nvazquez

File: agent/src/com/cloud/agent/Agent.java [hunk 21]
```diff
@@ -265,6 +273,8 @@ public void start() {
                 s_logger.info("Attempted to connect to the server, but received an unexpected exception, trying again...");
             }
         }
+        _shell.updateConnectedHost();
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: @nvazquez to be on safer/deterministic side, can this method accept the `host` String and set that instead of `hostToConnect`?

File: agent/src/com/cloud/agent/Agent.java [hunk 22]
```diff
@@ -699,16 +721,102 @@ public void processResponse(final Response response, final Link link) {
...
 
+        final List<String> msList = ready.getMsList();
+        if (msList != null && msList.size() > 0) {
+            final List<String> hosts = Arrays.asList(_shell.getHosts());
+            s_logger.info("Received management servers list: " + msList + ", current list: " + hosts);
+            if (isReceivedListUpdated(hosts, msList)) {
+                persistNewManagementServersList(msList);
+                _shell.resetHostCounter();
+                if (hasPrimaryHostChanged(hosts, msList)) {
+                    attemptReconnectionToPrimaryHost();
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: @nvazquez this can be removed, the ReadyCommand handler should not be responsible for reconnection/switching.

File: agent/src/com/cloud/agent/Agent.java [hunk 23]
```diff
@@ -699,16 +721,102 @@ public void processResponse(final Response response, final Link link) {
...
+        _shell.setHosts(newHosts);
+        _shell.setPersistentProperty(null, "host", newHosts);
+        s_logger.info("Saved new management servers list: " + msList);
+    }
+
+    /**
+     * Try reconnecting to preferred management server after new list is received.
+     * If connection is not performed, host counter is reset
+     */
+    private void attemptReconnectionToPrimaryHost() {
```
Comment (rhtyd) [inline agent/src/com/cloud/agent/Agent.java]: @nvazquez this method duplicates code and is not necessary, may be removed.

File: server/src/org/apache/cloudstack/agent/mslb/AgentMSLBServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+        List<String> msList = Arrays.asList(msServerAddresses.replace(" ", "").split(","));
+        AgentMSLBAlgorithm algorithm = getAgentMSLBAlgorithm();
+        return algorithm.getMSList(msList, orderedHostIds, hostId);
+    }
+
+    protected List<Long> getOrderedRunningHostIds(Long dcId) {
+        List<HostVO> hosts = resourceManager.listAllHostsInOneZoneByType(Host.Type.Routing, dcId);
+        if (hosts != null) {
+            return hosts.stream()
+                    .filter(x -> x.getHypervisorType().equals(Hypervisor.HypervisorType.KVM) &&
```
Comment (rhtyd) [inline server/src/org/apache/cloudstack/agent/mslb/AgentMSLBServiceImpl.java]: This should not be restricted just for KVM, but for any connected (indirect) agent.

File: agent/src/com/cloud/agent/Agent.java (hunk 21 above)
Comment (nvazquez) [reply agent/src/com/cloud/agent/Agent.java]: @rhtyd I decided to take that path as when agent is launched, the next host is set on constructor and then from `start()` method we couldn't know which host the agent is trying to connect to unless we do the same as in AgentShell with `hostToConnect` but on Agent class. Which approach do you think it would be best?

**Comment by nvazquez** [comment]
@rhtyd listeners for global settings update and new hosts added. Also shuffle is back to "normal" behavior. Lists will be propagated to agents only when 'hosts' or 'connected.agent.mslb.algorithm' values change or new hosts added, using new command-answer pattern instead of `ReadyCommand`. Please kindly review changes

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
just want to see jenkins succeed

**Comment by rhtyd** [comment]
Will open a new PR, closing this one.

**Comment by rhtyd** [comment]
New PR https://github.com/apache/cloudstack/pull/2469/files
